borsh-derive-internal/src/attribute_helpers.rs

```diff
@@ -18,3 +23,120 @@ pub fn contains_initialize_with(attrs: &[Attribute]) -> Option<Path> {
 
     None
 }
+pub fn check_use_discriminant(derive_input: DeriveInput) -> Result<Option<bool>, TokenStream> {
+    for attr in &derive_input.attrs {
+        if attr.path().is_ident("borsh") {
+            if let Meta::List(list) = attr.meta.clone() {
+                let tokens = list.tokens;
+                let meta: Meta = parse2(tokens).map_err(|err| err.to_compile_error())?;
+
+                if let Meta::NameValue(value) = meta {
+                    let MetaNameValue { path, value, .. } = value;
+                    if !path.is_ident(USE_DISCRIMINANT) {
+                        return Err(TokenStream::from(
+                            syn::Error::new(
+                                derive_input.ident.span(),
+                                "`use_discriminant` is the only supported attribute for `borsh`",
+                            )
+                            .to_compile_error(),
+                        ));
+                    }
+
+                    if let syn::Data::Struct(ref _data) = derive_input.data {
+                        return Err(TokenStream::from(
+                            syn::Error::new(
+                                derive_input.ident.span(),
+                                "borsh (use_discriminant=<bool>) does not support structs",
+                            )
+                            .to_compile_error(),
+                        ));
+                    }
+
+                    let value = value.to_token_stream().to_string();
+                    return match value.as_str() {
+                        "true" => Ok(Some(true)),
+                        "false" => Ok(Some(false)),
+                        _ => {
+                            return Err(TokenStream::from(
```

Comment:
More specifically , the silenced panic from `test_check_use_discriminant_wrong_value` happens on this line, when 
running `proc_macro::TokenStream::from` method in the `tests` context, i.e. outside of 
```
[lib]
proc-macro = true
```
context.